thinc/loss.py

```diff
@@ -32,92 +32,131 @@ def get_loss(self, guesses: GuessT, truths: TruthT) -> LossT:
 
 
 class CategoricalCrossentropy(Loss):
+    def __init__(self, *, normalize: bool = True):
+        self.normalize = normalize
+
     def __call__(self, guesses: Array2d, truths: Array) -> Tuple[Array2d, float]:
         return self.get_grad(guesses, truths), self.get_loss(guesses, truths)
 
     def get_grad(self, guesses: Array2d, truths: Array) -> Array2d:
         if truths.ndim != guesses.ndim:
+            # transform categorical values to one-hot encoding
             target = to_categorical(truths, n_classes=guesses.shape[-1])
         else:  # pragma: no cover
             target = cast(Array2d, truths)
         if guesses.shape != target.shape:  # pragma: no cover
-            err = f"Cannot calculate loss: mismatched shapes. {guesses.shape} vs {target.shape}"
+            err = f"Cannot calculate CategoricalCrossentropy loss: mismatched shapes: {guesses.shape} vs {target.shape}."
+            raise ValueError(err)
+        if guesses.any() > 1 or guesses.any() < 0:  # pragma: no cover
+            err = f"Cannot calculate CategoricalCrossentropy loss with guesses outside the [0,1] interval."
+            raise ValueError(err)
+        if target.any() > 1 or target.any() < 0:  # pragma: no cover
+            err = f"Cannot calculate CategoricalCrossentropy loss with truth values outside the [0,1] interval."
             raise ValueError(err)
         difference = guesses - target
-        return difference / guesses.shape[0]
+        if self.normalize:
+            difference = difference / guesses.shape[0]
+        return difference
 
     def get_loss(self, guesses: Array2d, truths: Array) -> float:
-        raise NotImplementedError
+        d_truth = self.get_grad(guesses, truths)
+        return (d_truth ** 2).sum()
 
 
 @registry.losses("CategoricalCrossentropy.v0")
-def configure_CategoricalCrossentropy() -> CategoricalCrossentropy:
-    return CategoricalCrossentropy()
+def configure_CategoricalCrossentropy(*, normalize: bool = True) -> CategoricalCrossentropy:
+    return CategoricalCrossentropy(normalize=normalize)
 
 
 class SequenceCategoricalCrossentropy(Loss):
-    def __init__(self):
-        self.cc = CategoricalCrossentropy()
+    def __init__(self, *, normalize: bool = True):
+        self.cc = CategoricalCrossentropy(normalize=normalize)
 
     def __call__(
         self, guesses: List[Array2d], truths: List[Array]
     ) -> Tuple[List[Array2d], List[float]]:
         return self.get_grad(guesses, truths), self.get_loss(guesses, truths)
 
     def get_grad(self, guesses: List[Array2d], truths: List[Array]) -> List[Array2d]:
-        if not guesses:
+        if guesses is None:
```

Comment:
I think we don't have to handle this sort of thing: our stance in Thinc is that the function behaves respectably if you give it respectable types. So we only need conditions for non-type invariants.